### PR TITLE
glob in Windows breaks document-watch

### LINKED ISSUE #2732: Search does not work on Windows local build
This occurs on Windows for all local builds, and has done for some weeks (previously was excellent).

The error is "Error initializing search index" - see below.
image

### LINKED ISSUE #2733: What do we do about all the slug warnings
On current main, when I start yari I get a huge number of messages like:
```
The slug of files/en-us/games/anatomy/index.html doesn't match the folder is located in.
```
Is this an error I should do something about? If so, what?
If not, should this be logged somewhere else?

### PR DESCRIPTION
**Pull request by peterbe**
Fixes #2733
Fixes #2732

@hamishwillee If you have the time, do you think you can test this?
You'd have to have a git checkout of this branch of mine of the `yari` repo. You can't test this from the `mdn/content` repo yet. 

I don't have access to Windows to test, but I'm fairly confident it will work. But your functional test-blessing would be great.

### CODE REVIEW

**Comment by hamishwillee** [comment]
Hi @peterbe 
I am happy to test, and I know how to pull branches. What I don't know how to do is pull this branch onto my content repo so that I am using this instead of whatever mdn/content uses by default. 
Advice?

**Comment by peterbe** [comment]
> Hi @peterbe 
> 
> I am happy to test, and I know how to pull branches. What I don't know how to do is pull this branch onto my content repo so that I am using this instead of whatever mdn/content uses by default. 
> 
> Advice?

You'll need to test it all from the mdn/yari repo. Can't use the mdn/content one. 

Alternatively you can just edit the same file as changed in this PR but inside the node_modules over in your mdn/content repo. 

Mind you, we can just merge this and see if it just fixes itself by early next week.

**Comment by Ryuno-Ki** [comment]
> What I don't know how to do is pull this branch onto my content repo so that I am using this instead of whatever mdn/content uses by default.
> Advice?

This should work:

```
git remote add peterbe https://github.com/peterbe/yari.git
git fetch peterbe
git checkout peterbe/2733-glob-in-windows-breaks-document-watch
```
(OTOH)

**Comment by mattiapontonio** [comment]
> > What I don't know how to do is pull this branch onto my content repo so that I am using this instead of whatever mdn/content uses by default.
> > Advice?
> 
> This should work:
> 
> ```
> git remote add peterbe https://github.com/peterbe/yari.git
> git fetch peterbe
> git checkout peterbe/2733-glob-in-windows-breaks-document-watch
> ```
> (OTOH)

That's good 👍

Then
```
npm i
npm start
```
?

Readme.

**Comment by hamishwillee** [comment]
Thanks guys. My problem wasn't how to get a branch (I prefer PR pulls for testing `git fetch origin pull/2748/head:testsearch`) but how to run yari independently. Found instructions in readme. Duh! But those instructions fail at `yarn dev` on Windows:
```
$ cross-env NODE_ENV=production node build/cli.js --spas-only

Building SPAs...
Wrote D:\github\mozilla\mdn_content\yari\client\build\en-us\_spas\404.html
$ node tool/cli.js gather-git-history --verbose
Error: ENOENT: no such file or directory, open '..\content\files\en-us\api\file_api\index.html\_githistory.json'
    at Object.openSync (fs.js:458:3)
    at Object.writeFileSync (fs.js:1355:35)
    at D:\github\mozilla\mdn_content\yari\tool\cli.js:318:12
    at Se._action (D:\github\mozilla\mdn_content\yari\tool\cli.js:27:13)
    at Se.run (D:\github\mozilla\mdn_content\yari\node_modules\@caporal\core\dist\index.js:1:27590)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
    at async Te._run (D:\github\mozilla\mdn_content\yari\node_modules\@caporal\core\dist\index.js:1:32257)

error: ENOENT: no such file or directory, open '..\content\files\en-us\api\file_api\index.html\_githistory.json'

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
...
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

@peterbe I'm happy enough for you to just merge.

File: server/document-watch.worker.js [hunk 1]
```diff
@@ -26,7 +26,15 @@ function postDocumentInfo(filePath, changeType) {
...
     // This would prevent the document from being added in the first place
     // if the filePath doesn't map correctly to the URL, but in reverse.
-    if (!filePath.includes(Document.urlToFolderPath(document.url))) {
+    // Note! On Windows, `glob.sync()` will return URLs that always use
+    // the `/` character (POSIX STYLE, https://www.npmjs.com/package/glob#windows)
+    // But the `Document.urlToFolderPath()` function will respect Windows
+    // notation and use `\` characters.
+    if (
+      !filePath.includes(
+        Document.urlToFolderPath(document.url).replace(/\\/g, "/")
```
Comment (fiji-flo) [inline server/document-watch.worker.js:35]: I'd suggest we transform `filePath` before we invoke `postDocumentInfo`. I'm curious why the `Document.read` works in this case it it does. Isn't `filePath` something like `en-us/web/css/index.html` on a windows system but `Document.read` expects `en-us\web\css\index.html` ?

I can test it on windows with some effort, should I?

Comment (peterbe) [reply server/document-watch.worker.js:35]: So, you suggest the other way around. I think I get it. That way, the `filePath` gets Windows'ified onwards. I'll look into it.

File: server/document-watch.worker.js [hunk 2]
```diff
@@ -14,6 +14,14 @@ function postEvent(type, data = {}) {
 }
 
 function postDocumentInfo(filePath, changeType) {
+  // Note! On Windows, `glob.sync()` will return URLs that always use
+  // the `/` character (POSIX STYLE, https://www.npmjs.com/package/glob#windows)
+  // But the `Document.urlToFolderPath()` function will respect Windows
+  // notation and use `\` characters.
+  // Now, when we use it, it'll work as if `glob.sync()` and returned paths
+  // in a fashion that is expected in the OS.
+  filePath = path.split(filePath).join(path.sep);
```
Comment (peterbe) [inline server/document-watch.worker.js:23]: What do you think about this solution @fiji-flo ?

Comment (fiji-flo) [reply server/document-watch.worker.js]: I think you need and explicit `path.join(filePath.split("/"))`

I assume `path.split` splits only on `path.sep` which would be `\` in this case but we wanna split `filePath` on `/`, don't we?

Comment (fiji-flo) [reply server/document-watch.worker.js]: Did you literally push my proposed change while I was writing this? Spooky

File: server/document-watch.worker.js [hunk 3]
```diff
@@ -14,6 +14,14 @@ function postEvent(type, data = {}) {
 }
 
 function postDocumentInfo(filePath, changeType) {
+  // Note! On Windows, `glob.sync()` will return URLs that always use
+  // the `/` character (POSIX STYLE, https://www.npmjs.com/package/glob#windows)
+  // But the `Document.urlToFolderPath()` function will respect Windows
+  // notation and use `\` characters.
+  // Now, when we use it, it'll work as if `glob.sync()` and returned paths
+  // in a fashion that is expected in the OS.
+  filePath = filePath.split("/").join(path.sep);
```
Comment (fiji-flo) [inline server/document-watch.worker.js:23]: I would move this to line 56 https://github.com/mdn/yari/pull/2748/files#diff-7c825b83e0a9311cc1fe49ed6075267dbcd2c0aad8c61a008d178909d2633f07R56

**Comment by peterbe** [comment]
ping @fiji-flo

**Review by fiji-flo — APPROVED** [review]
👍

**Comment by peterbe** [comment]
@hamishwillee In a little while, update your `main` and `yarn install` and try again. Please. /me hoping

**Comment by hamishwillee** [comment]
Awesome. Fixes both bugs.

**Pull request merged**